File: examples/python/nemotron_speech.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

import argparse
import importlib
import json
import os
import sys
import time

import numpy as np
import onnxruntime_genai as og
from common import get_config

# Maps short language codes / locale tags to (lang_id, human-readable name).
# Restricted to the languages officially supported by
# NVIDIA-Nemotron-3.5-ASR-Streaming-Multilingual-0.6b (per model card).
# IDs follow the canonical prompt_dictionary in model_config.yaml.
LANG_TO_ID = {
    "en": (0, "English (default / US)"),
    "en-US": (0, "English (United States)"),
    "en-GB": (1, "English (United Kingdom)"),
    "es-ES": (2, "Spanish (Spain)"),
    "es": (3, "Spanish (default / Latin America)"),
    "es-US": (3, "Spanish (US Latin American)"),
    "zh-CN": (4, "Chinese (Mandarin, Simplified)"),
    "hi": (6, "Hindi"),
    "hi-IN": (6, "Hindi (India)"),
    "ar": (7, "Arabic"),
    "ar-AR": (7, "Arabic"),
    "fr": (8, "French (default / France)"),
    "fr-FR": (8, "French (France)"),
    "de": (9, "German"),
    "de-DE": (9, "German (Germany)"),
    "ja": (10, "Japanese"),
    "ja-JP": (10, "Japanese"),
    "ru": (11, "Russian"),
    "ru-RU": (11, "Russian"),
    "pt-BR": (12, "Portuguese (Brazil)"),
    "pt": (13, "Portuguese (default / Portugal)"),
    "pt-PT": (13, "Portuguese (Portugal)"),
    "ko": (14, "Korean"),
    "ko-KR": (14, "Korean (South Korea)"),
    "it": (15, "Italian"),
    "it-IT": (15, "Italian"),
    "nl": (16, "Dutch"),
    "nl-NL": (16, "Dutch (Netherlands)"),
    "pl": (17, "Polish"),
    "pl-PL": (17, "Polish"),
    "tr": (18, "Turkish"),
    "tr-TR": (18, "Turkish"),
    "uk": (19, "Ukrainian"),
    "uk-UA": (19, "Ukrainian"),
    "ro": (20, "Romanian"),
    "ro-RO": (20, "Romanian"),
    "el": (21, "Greek"),
    "el-GR": (21, "Greek"),
    "cs": (22, "Czech"),
    "cs-CZ": (22, "Czech"),
    "hu": (23, "Hungarian"),
    "hu-HU": (23, "Hungarian"),
    "sv": (24, "Swedish"),
    "sv-SE": (24, "Swedish"),
    "da": (25, "Danish"),
    "da-DK": (25, "Danish"),
    "fi": (26, "Finnish"),
    "fi-FI": (26, "Finnish"),
    "sk": (28, "Slovak"),
    "sk-SK": (28, "Slovak"),
    "hr": (29, "Croatian"),
    "hr-HR": (29, "Croatian"),
    "bg": (30, "Bulgarian"),
    "bg-BG": (30, "Bulgarian"),
    "lt": (31, "Lithuanian"),
    "lt-LT": (31, "Lithuanian"),
    "th": (32, "Thai"),
    "th-TH": (32, "Thai"),
    "vi": (33, "Vietnamese"),
    "vi-VN": (33, "Vietnamese"),
    "et": (60, "Estonian"),
    "et-EE": (60, "Estonian"),
    "lv": (61, "Latvian"),
    "lv-LV": (61, "Latvian"),
    "sl": (62, "Slovenian"),
    "sl-SI": (62, "Slovenian"),
    "he": (64, "Hebrew"),
    "he-IL": (64, "Hebrew (Israel)"),
    "fr-CA": (100, "French (Canada)"),
    "auto": (101, "Auto-detect"),
    "mt": (102, "Maltese"),
    "mt-MT": (102, "Maltese"),
    "nb": (103, "Norwegian Bokmål"),
    "nb-NO": (103, "Norwegian Bokmål"),
    "nn": (104, "Norwegian Nynorsk"),
    "nn-NO": (104, "Norwegian Nynorsk"),
}


def load_config(model_path):
    """Read sample_rate and chunk_samples from genai_config.json."""
    config_path = os.path.join(model_path, "genai_config.json")
    with open(config_path) as f:
        config = json.load(f)
    sample_rate = config["model"]["sample_rate"]
    chunk_samples = config["model"]["chunk_samples"]
    return sample_rate, chunk_samples


def load_audio(audio_path, sample_rate):
    sf = importlib.import_module("soundfile")

    audio, sr = sf.read(audio_path, dtype="float32")
    if len(audio.shape) > 1:
        audio = audio.mean(axis=1)
    if sr != sample_rate:
        scipy_signal = importlib.import_module("scipy.signal")

        num_samples = int(len(audio) * sample_rate / sr)
        audio = scipy_signal.resample(audio, num_samples).astype(np.float32)
    return audio


def decode_tokens(generator, tokenizer_stream):
    """Decode all available tokens from the generator, returning the text."""
    text = ""
    while not generator.is_done():
        generator.generate_next_token()
        tokens = generator.get_next_tokens()
        if len(tokens) > 0:
            token_text = tokenizer_stream.decode(tokens[0])
            if token_text:
                print(token_text, end="", flush=True)
                text += token_text
    return text


def simulate_microphone(model_path, audio_path, execution_provider, use_vad=None, language=None):
    """Stream audio through Generator + StreamingProcessor API."""
    sample_rate, chunk_samples = load_config(model_path)
    audio = load_audio(audio_path, sample_rate)
    duration = len(audio) / sample_rate

    config = get_config(model_path, execution_provider, None)
    selected_lang = None
    if language is not None:
        if language not in LANG_TO_ID:
            raise ValueError(f"Unknown language '{language}'. Known: {sorted(LANG_TO_ID)}")
        selected_lang = LANG_TO_ID[language]
        lang_id, lang_name = selected_lang
        print(f"  Language: {language} -> {lang_name} (lang_id={lang_id})")
    model = og.Model(config)
    processor = og.StreamingProcessor(model)

    # VAD is off by default. Use --use_vad true to enable (requires "vad" section in genai_config.json).
    processor.set_option("use_vad", "false")
    if use_vad:
        try:
            processor.set_option("use_vad", "true")
        except Exception as e:
            print(f"  VAD: disabled (no VAD config in genai_config.json: {e})")
    vad_status = processor.get_option("use_vad")
    print(f"  Use VAD: {vad_status}")
    if vad_status == "true":
        print(f"  VAD threshold: {processor.get_option('vad_threshold')}")

    tokenizer = og.Tokenizer(model)
    tokenizer_stream = tokenizer.create_stream()
    params = og.GeneratorParams(model)
    generator = og.Generator(model, params)
    # Per-generator language selection
    if selected_lang is not None:
        generator.set_runtime_option("lang_id", str(int(selected_lang[0])))

    print("-" * 60)
    stream_start = time.perf_counter()
    full_transcript = ""
    vad_enabled = vad_status == "true"
    chunks_total = 0
    chunks_processed = 0
    chunks_skipped = 0

    for i in range(0, len(audio), chunk_samples):
        chunk = audio[i : i + chunk_samples].astype(np.float32)
        chunks_total += 1
        inputs = processor.process(chunk)
        if inputs is not None:
            chunks_processed += 1
            generator.set_inputs(inputs)
            full_transcript += decode_tokens(generator, tokenizer_stream)
        else:
            chunks_skipped += 1

    # Flush remaining audio
    inputs = processor.flush()
    if inputs is not None:
        generator.set_inputs(inputs)
        full_transcript += decode_tokens(generator, tokenizer_stream)

    total_wall = time.perf_counter() - stream_start

    print(f"\n{'=' * 60}")
    print(f"  {full_transcript.strip()}")
    print(f"{'=' * 60}")
    print(f"  Audio: {duration:.2f}s | Wall: {total_wall:.2f}s | RTF: {duration / total_wall:.2f}x")
    if vad_enabled:
        pct_saved = chunks_skipped / max(chunks_total, 1) * 100
        print(
            f"  VAD Metrics: {chunks_total} total chunks, {chunks_processed} processed, "
            f"{chunks_skipped} skipped ({pct_saved:.1f}% compute saved)"
        )


def main():
    parser = argparse.ArgumentParser(formatter_class=argparse.RawTextHelpFormatter)
    parser.add_argument("--model_path", type=str, required=True)
    parser.add_argument("--audio_file", type=str, required=True)
    parser.add_argument(
        "--use_vad",
        type=str,
        choices=["true", "false"],
        default=None,
        help="Override VAD setting from genai_config.json (true/false).",
    )
    lang_help = (
        "Language / locale code for the multilingual encoder. "
        "Overrides model.default_lang_id from genai_config.json. "
        "Use a bare ISO 639-1 code (e.g. 'de', 'fr', 'pt') for the default locale, "
        "or a BCP-47 locale tag for region-specific variants. "
        "Pass 'auto' to let the model detect the language. Supported codes:\n"
        + "\n".join(f"  {code:<7} {name}" for code, (_, name) in sorted(LANG_TO_ID.items()))
    )
    parser.add_argument("--language", "-l", type=str, default=None, help=lang_help)
    parser.add_argument(
        "-e",
        "--execution_provider",
        type=str,
        required=False,
        default="follow_config",
        choices=["cpu", "cuda", "dml", "follow_config"],
        help="Execution provider to run with. Defaults to follow_config.",
    )
    args = parser.parse_args()
    if not os.path.exists(args.audio_file):
        print(f"Error: {args.audio_file} not found")
        sys.exit(1)
    use_vad_override = None
    if args.use_vad is not None:
        use_vad_override = args.use_vad == "true"
    simulate_microphone(
        args.model_path, args.audio_file, args.execution_provider, use_vad=use_vad_override, language=args.language
    )


if __name__ == "__main__":
    main()

File: examples/c/src/nemotron_speech.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
//
// nemotron_speech.cpp — Streaming ASR example using StreamingProcessor + Generator API.
//
// Usage:
//   ./nemotron_speech --model_path /path/to/nemotron-model --audio_file /path/to/audio.wav

#include <chrono>
#include <cstring>
#include <fstream>
#include <iostream>
#include <string>
#include <vector>

#include <nlohmann/json.hpp>
#include "ort_genai.h"

struct AudioConfig {
  int sample_rate;
  int chunk_samples;
};

AudioConfig LoadConfig(const std::string& model_path) {
  std::string config_path = model_path + "/genai_config.json";
  std::ifstream f(config_path);
  if (!f.is_open()) {
    throw std::runtime_error("Cannot open " + config_path);
  }
  auto config = nlohmann::json::parse(f);
  return {
      config["model"]["sample_rate"].get<int>(),
      config["model"]["chunk_samples"].get<int>(),
  };
}

// Simple WAV loader — expects 16-bit PCM, mono or stereo.
// Returns float32 samples normalized to [-1, 1].
std::vector<float> LoadWav(const std::string& path, int target_sample_rate) {
  std::ifstream file(path, std::ios::binary);
  if (!file.is_open()) {
    throw std::runtime_error("Cannot open audio file: " + path);
  }

  // Read WAV header
  char riff[4];
  file.read(riff, 4);
  if (std::memcmp(riff, "RIFF", 4) != 0) {
    throw std::runtime_error("Not a valid WAV file (missing RIFF header)");
  }

  file.seekg(4, std::ios::cur);  // Skip file size

  char wave[4];
  file.read(wave, 4);
  if (std::memcmp(wave, "WAVE", 4) != 0) {
    throw std::runtime_error("Not a valid WAV file (missing WAVE marker)");
  }

  // Find fmt chunk
  int16_t num_channels = 0;
  int32_t sample_rate = 0;
  int16_t bits_per_sample = 0;

  while (file.good()) {
    char chunk_id[4];
    int32_t chunk_size;
    file.read(chunk_id, 4);
    file.read(reinterpret_cast<char*>(&chunk_size), 4);

    if (std::memcmp(chunk_id, "fmt ", 4) == 0) {
      int16_t audio_format;
      file.read(reinterpret_cast<char*>(&audio_format), 2);
      file.read(reinterpret_cast<char*>(&num_channels), 2);
      file.read(reinterpret_cast<char*>(&sample_rate), 4);
      file.seekg(6, std::ios::cur);  // Skip byte rate + block align
      file.read(reinterpret_cast<char*>(&bits_per_sample), 2);
      if (chunk_size > 16) {
        file.seekg(chunk_size - 16, std::ios::cur);
      }
    } else if (std::memcmp(chunk_id, "data", 4) == 0) {
      int num_samples = chunk_size / (bits_per_sample / 8) / num_channels;
      std::vector<float> audio(num_samples);

      if (bits_per_sample == 16) {
        std::vector<int16_t> raw(num_samples * num_channels);
        file.read(reinterpret_cast<char*>(raw.data()), chunk_size);
        for (int i = 0; i < num_samples; i++) {
          if (num_channels == 1) {
            audio[i] = raw[i] / 32768.0f;
          } else {
            // Average channels
            float sum = 0.0f;
            for (int c = 0; c < num_channels; c++) {
              sum += raw[i * num_channels + c];
            }
            audio[i] = (sum / num_channels) / 32768.0f;
          }
        }
      } else if (bits_per_sample == 32) {
        // Assume float32
        std::vector<float> raw(num_samples * num_channels);
        file.read(reinterpret_cast<char*>(raw.data()), chunk_size);
        for (int i = 0; i < num_samples; i++) {
          if (num_channels == 1) {
            audio[i] = raw[i];
          } else {
            float sum = 0.0f;
            for (int c = 0; c < num_channels; c++) {
              sum += raw[i * num_channels + c];
            }
            audio[i] = sum / num_channels;
          }
        }
      } else {
        throw std::runtime_error("Unsupported bits per sample: " + std::to_string(bits_per_sample));
      }

      // Basic resampling if needed (linear interpolation)
      if (sample_rate != target_sample_rate) {
        int new_len = static_cast<int>(audio.size() * static_cast<double>(target_sample_rate) / sample_rate);
        std::vector<float> resampled(new_len);
        for (int i = 0; i < new_len; i++) {
          double src_idx = i * static_cast<double>(audio.size() - 1) / (new_len - 1);
          int idx0 = static_cast<int>(src_idx);
          int idx1 = std::min(idx0 + 1, static_cast<int>(audio.size()) - 1);
          double frac = src_idx - idx0;
          resampled[i] = static_cast<float>(audio[idx0] * (1.0 - frac) + audio[idx1] * frac);
        }
        return resampled;
      }

      return audio;
    } else {
      file.seekg(chunk_size, std::ios::cur);
    }
  }

  throw std::runtime_error("No data chunk found in WAV file");
}

std::string DecodeTokens(OgaGenerator& generator, OgaTokenizerStream& tokenizer_stream) {
  std::string text;
  while (!generator.IsDone()) {
    generator.GenerateNextToken();
    auto next_tokens = generator.GetNextTokens();
    if (!next_tokens.empty()) {
      const char* token_text = tokenizer_stream.Decode(next_tokens[0]);
      if (token_text && token_text[0] != '\0') {
        std::cout << token_text << std::flush;
        text += token_text;
      }
    }
  }
  return text;
}

void StreamingTranscribe(const std::string& model_path, const std::string& audio_path, bool enable_vad = false) {
  auto [sample_rate, chunk_samples] = LoadConfig(model_path);

  std::cout << "Loading audio: " << audio_path << std::endl;
  auto audio = LoadWav(audio_path, sample_rate);
  double duration = static_cast<double>(audio.size()) / sample_rate;

  std::cout << "Loading model: " << model_path << std::endl;
  auto config = OgaConfig::Create(model_path.c_str());
  auto model = OgaModel::Create(*config);
  auto processor = OgaStreamingProcessor::Create(*model);

  // VAD is disabled by default. Enable via genai_config.json or --enable_vad.
  if (enable_vad && std::string(processor->GetOption("vad_enabled")) != "true") {
    processor->SetOption("vad_enabled", "true");
  }
  if (std::string(processor->GetOption("vad_enabled")) == "true") {
    std::cout << "  VAD: enabled" << std::endl;
  } else {
    std::cout << "  VAD: disabled" << std::endl;
  }

  auto tokenizer = OgaTokenizer::Create(*model);
  auto tokenizer_stream = OgaTokenizerStream::Create(*tokenizer);
  auto params = OgaGeneratorParams::Create(*model);
  auto generator = OgaGenerator::Create(*model, *params);

  std::cout << "  Sample rate: " << sample_rate << ", Chunk: " << chunk_samples << " samples" << std::endl;
  std::cout << "  Audio duration: " << duration << "s" << std::endl;
  std::cout << std::string(60, '-') << std::endl;

  auto start = std::chrono::high_resolution_clock::now();
  std::string full_transcript;

  // Stream audio in chunks
  for (size_t i = 0; i < audio.size(); i += chunk_samples) {
    size_t remaining = std::min(static_cast<size_t>(chunk_samples), audio.size() - i);
    auto inputs = processor->Process(audio.data() + i, remaining);
    if (inputs) {
      generator->SetInputs(*inputs);
      full_transcript += DecodeTokens(*generator, *tokenizer_stream);
    }
  }

  // Flush remaining audio
  {
    auto inputs = processor->Flush();
    if (inputs && inputs.get()) {
      generator->SetInputs(*inputs);
      full_transcript += DecodeTokens(*generator, *tokenizer_stream);
    }
  }

  auto end = std::chrono::high_resolution_clock::now();
  double wall_time = std::chrono::duration<double>(end - start).count();

  std::cout << "\n"
            << std::string(60, '=') << std::endl;
  std::cout << "  " << full_transcript << std::endl;
  std::cout << std::string(60, '=') << std::endl;
  std::cout << "  Audio: " << duration << "s | Wall: " << wall_time << "s | RTF: " << (duration / wall_time) << "x" << std::endl;
}

int main(int argc, char* argv[]) {
  if (argc < 3) {
    std::cerr << "Usage: " << argv[0] << " --model_path <path> --audio_file <path>" << std::endl;
    return 1;
  }

  std::string model_path;
  std::string audio_file;
  bool enable_vad = false;

  for (int i = 1; i < argc; i++) {
    if (std::string(argv[i]) == "--model_path" && i + 1 < argc) {
      model_path = argv[++i];
    } else if (std::string(argv[i]) == "--audio_file" && i + 1 < argc) {
      audio_file = argv[++i];
    } else if (std::string(argv[i]) == "--enable_vad") {
      enable_vad = true;
    }
  }

  if (model_path.empty() || audio_file.empty()) {
    std::cerr << "Both --model_path and --audio_file are required." << std::endl;
    return 1;
  }

  try {
    StreamingTranscribe(model_path, audio_file, enable_vad);
  } catch (const std::exception& e) {
    std::cerr << "Error: " << e.what() << std::endl;
    return 1;
  }

  return 0;
}
